### DISCUSSION TITLE
CacheProvider not used in production

### DISCUSSION THREAD

**Discussion by sarink** [emotion-js/emotion#2903] [Q&A] [answered]
**Current behavior:**

We are using a custom `cache` and `CacheProvider` in our client-side rendered SPA so that `<style>` tags can be injected into the `<head>` (because the default implementation is to use `insertRule`).

In development mode, this works great, I see zillions of `<style>` tags in the head. However, when deploying to production, I see one empty style tag in the dom: `<style data-emotion="css" data-s=""></style>`. It's as if our `CacheProvider` is being completely ignored in production.

**To reproduce:**

```
const muiTheme = createTheme();

const cache = createCache({ key: 'css', container: document.head });

export const AppContainer: React.FC = () => (
    <CacheProvider value={cache}>
      <ThemeProvider theme={muiTheme}>
        <App/>
      </ThemeProvider>
    </CacheProvider>
);
```

1. Do a build with `NODE_ENV = "development"`, observe `<style>` tags in the head
2. Do a build with `NODE_ENV = "production"`, observe there are no `<style>` tags in the head

**Expected behavior:**

I expect the `CacheProvider` to work in production the same way it does in development

**Environment information:**

- `react` version: 17.0.2
- `@emotion/react` version: 11.9.3

**Comment by srmagura** [comment, marked as answer]
In production, `@emotion/sheet` uses "speedy mode" which puts up to 65,000 style rules in a single style tag. They are inserted directly via the CSSOM rather than adding the CSS text to the inside of the `<style>` tag.

Speedy mode isn't as convenient for debugging style issues which is why it's not enabled by default in development. 

If you want to disable speedy mode all together, you can do `createCache({ key: 'sarink', speedy: false })`. This will decrease performance of course so I would not recommend it.

**Reply by ManoMangaldas** [reply]
Thank you.

I had issues in production where randomly, the class that was supposed to be present via the insertRule API was missing. The MUI control referenced a class, but not applied. Guess it was too speedy.

**Comment by sarink** [comment]
Interesting, that fixed it, thank you @srmagura! Might I suggest some docs on `speedy` mode? I don't see this mentioned anywhere on the CacheProvider page.

Also, as prerendering is still a popular way for SPAs to satisfy SEO requirements, styles in the head are necessary for this, and many folks familiar with traditional css-in-js are probably accustomed to styles being injected into the head, this new-ish implementation comes as quite a surprise.

I completely understand that `insertRule` is faster and offers performance benefits, and am thankful that there's a `speedy` flag to toggle it (we will be using this only during the prerender phase of our build), but I think this ought to be documented a little more clearly.

Thanks!

**Reply by srmagura** [reply]
A documentation PR for this would be welcome! This is the file you would want to update: https://github.com/emotion-js/emotion/blob/main/packages/cache/README.md.

Hmm, if you are prerendering, Emotion should use the SSR code path. In our automatic SSR support we insert `<style>` tags (with actual text content) into the `<body>`. https://emotion.sh/docs/ssr     

I assume your prerendering occurs within a typical Node environment? Are you using a framework e.g. Next.js?

**Reply by sarink** [reply]
No, we are prerendering a client-side-rendered SPA (a la prerender.io) - these kinds of tools work by navigating to the page, waiting for javascript to finish executing, then exporting the contents of the dom to an html file. It's a cheap lightweight way to achieve similar SEO capabilities without a server at all.

**Reply by Andarist** [reply]
The speedy thing + prerendering is already mentioned here: https://emotion.sh/docs/ssr#puppeteer 

Maybe those information could be moved elsewhere though

**Reply by sarink** [reply]
👀 I guess I missed it. I do think that `speedy` should be mentioned on the `CacheProvider` page as well.

Thanks for your quick responses and helping me find this option!

**Reply by fpmanuel** [reply]
@srmagura Speedy hasn't been documented yet, right? I can add documentation for it if needed.

**Comment by erikjalevik** [comment]
I'm using the `CacheProvider` to inject styles into a second window opened via `window.open` (as described [here](https://github.com/emotion-js/emotion/issues/1229)).

This also breaks in production, the styles are missing from the opened window. It seems I can fix it as suggested above by setting `speedy: false` when creating the cache for the new window. However, I was wondering if there's a way to handle this without disabling speedy mode?

**Reply by Andarist** [reply]
It should "just work". Speedy mode just changes how the styles are inserted - it doesn't affect what is inserted or removed. Please prepare a repro case for your problem to get a further help

**Reply by erikjalevik** [reply]
Unfortunately it doesn't just work. :) It's in an Electron app, if that makes a difference.

I'll see if I can narrow it down to a shareable repro case.

**Reply by erikjalevik** [reply]
Actually, all that's needed to reproduce is to add a `speedy: true` to the `createCache` statement in the sandbox given as an example in the other issue:

https://codesandbox.io/s/silly-burnell-c4ncs?file=/src/App.tsx

The problem seems to be that `sheetForTag` inside `@emotion/sheet` returns null when a cache is created with a detached element like `containerEl` here.

**Reply by Andarist** [reply]
I see - yeah, this might not work fully with detached containers. In such a situation you should not use the speedy mode. Perhaps we could add some kind of a dev-only warning to raise this as a potential issue.

**Reply by ManoMangaldas** [reply]
> It should "just work". Speedy mode just changes how the styles are inserted - it doesn't affect what is inserted or removed. Please prepare a repro case for your problem to get a further help

In my case, application is hosted Azure Web Apps as a SPA. It isnt working consistently. speedy: false is consistent.
`<AppRouterCacheProvider key={"css"} options={{ speedy: false }}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </AppRouterCacheProvider>`

**Comment by sashless** [comment]
For some reasons `speedy: false`  on `cache` is not respected when doing a `static storybook` build.

In `live storybook` this works.

I am using `@storybook: 7.6.15` and tried `@emotion/cache: 10.9.3` and `@emotion/cache: 11.11.0` without luck.

**Reply by Andarist** [reply]
Without a repro, I can't investigate this.

**Reply by sashless** [reply]
Yea you are right. For now i am in touch with support @ chromatic regarding storybooks issues there. When there is no trace for a solution i will create a repro. And when we find a solution i will update here.  At first i was only in hope for some kind of hint. I should have mention the whole story before. Sorry for that @Andarist

**Reply by sashless** [reply]
The storybook static build was forcing a `production` build. It could not be set via webpacks config and needed to be set via `NODE_ENV` as in `NODE_ENV=development build-storybook`. With this `speedy` flag is respected.
